### ISSUE TITLE
Error: The element reference of <button class="_5f5mN       jIbKX  _6VtSN     yZn4P   "> is stale

### ISSUE THREAD

**Issue by bradwalls** [InstaPy/InstaPy#5039]
## Current Behavior

```
Traceback (most recent call last):
  File "bc.py", line 63, in <module>
    session.interact_user_followers(['emjeiiamira', 'aquabumps', 'josh_burkinshaw', 'scottydman'], amount=50, randomize=True)
  File "/home/pi/.local/lib/python3.7/site-packages/instapy/instapy.py", line 3185, in interact_user_followers
    self.user_interact_media,
  File "/home/pi/.local/lib/python3.7/site-packages/instapy/instapy.py", line 2471, in interact_by_users
    media,
  File "/home/pi/.local/lib/python3.7/site-packages/instapy/like_util.py", line 450, in get_links_for_username
    browser, "profile", username, person, None, logger, logfolder
  File "/home/pi/.local/lib/python3.7/site-packages/instapy/unfollow_util.py", line 211, in get_following_status
    following_status = follow_button.text
  File "/home/pi/.local/lib/python3.7/site-packages/selenium/webdriver/remote/webelement.py", line 76, in text
    return self._execute(Command.GET_ELEMENT_TEXT)['value']
  File "/home/pi/.local/lib/python3.7/site-packages/selenium/webdriver/remote/webelement.py", line 633, in _execute
    return self._parent.execute(command, params)
  File "/home/pi/.local/lib/python3.7/site-packages/selenium/webdriver/remote/webdriver.py", line 321, in execute
    self.error_handler.check_response(response)
  File "/home/pi/.local/lib/python3.7/site-packages/selenium/webdriver/remote/errorhandler.py", line 242, in check_response
    raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.StaleElementReferenceException: Message: The element reference of <button class="_5f5mN       jIbKX  _6VtSN     yZn4P   "> is stale; either the element is no longer attached to the DOM, it is not in the current frame context, or the document has been refreshed
```

## InstaPy configuration
```
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import random, time, traceback
from instapy import InstaPy
from instapy import smart_run
from instapy import set_workspace

# set workspace folder at desired location (default is at your home folder)
set_workspace(path=None)

# login credentials
insta_username = 'Bradscanvas'
insta_password = 'password'

# restriction data


session = InstaPy(username=insta_username,
                  password=insta_password,
                  headless_browser=False,
                  disable_image_load=True)
                  
with smart_run(session):

        session.set_mandatory_language(enabled=True, character_set=['LATIN'])        
        session.set_quota_supervisor(enabled=True, sleep_after=["follows", "server_calls"],
                                     sleepyhead=True,
                                     stochastic_flow=True,
                                     notify_me=True,
                                     peak_follows_hourly=18,
                                     peak_follows_daily=None,
                                     peak_unfollows_hourly=26,
                                     peak_unfollows_daily=None,
                                     peak_server_calls_hourly=188,
                                     peak_server_calls_daily=4200)
        
        session.set_simulation(enabled=True, percentage=73)
        session.set_action_delays(enabled=True,
                                  follow=16.88,
                                  unfollow=14.91,
                                  randomize=True,
                                  random_range_from=214,
                                  random_range_to=447)
                
        session.set_relationship_bounds(enabled=True,
                                        potency_ratio=1.12,
                                        delimit_by_numbers=True,
                                        max_followers=46880,
                                        max_following=26880,
                                        max_posts=1000,
                                        min_posts=6,
                                        min_followers=6,
                                        min_following=6)
        
        session.set_skip_users(skip_private=True, private_percentage=74,
                               skip_no_profile_pic=True, no_profile_pic_percentage=100,
                               skip_business=True)

        session.set_user_interact(amount=5, randomize=True, percentage=50, media='Photo')
        session.set_do_follow(enabled=True, percentage=70)
        session.set_do_like(enabled=True, percentage=100)
        session.interact_user_followers(['emjeiiamira', 'aquabumps', 'josh_burkinshaw', 'scottydman'], amount=50, randomize=True)
        

```

**Comment by alexsmedile** [comment]
Same here
```
Message: The element reference of <button class="_5f5mN       jIbKX  _6VtSN     yZn4P   ">
is stale; either the element is no longer attached to the DOM, it is not in the current
frame context, or the document has been refreshed.
```

**Comment by alexsmedile** [comment]
Possible fix: #5038

**Comment by Jostino** [comment]
same error:
```
Traceback (most recent call last):
  File "/root/InstaPy/assets/freemake.py", line 33, in <module>
    session.set_dont_unfollow_active_users(enabled=True, posts=3)
  File "/root/.local/lib/python3.7/site-packages/instapy/instapy.py", line 4225, in set_dont_unfollow_active_users
    self.browser, self.username, posts, boundary, self.logger
  File "/root/.local/lib/python3.7/site-packages/instapy/util.py", line 762, in get_active_users
    user_list = get_users_from_dialog(user_list, dialog)
  File "/root/.local/lib/python3.7/site-packages/instapy/util.py", line 2133, in get_users_from_dialog
    user_blocks = dialog.find_elements_by_tag_name("a")
  File "/root/.local/lib/python3.7/site-packages/selenium/webdriver/remote/webelement.py", line 320, in find_elements_by_tag_name
    return self.find_elements(by=By.TAG_NAME, value=name)
  File "/root/.local/lib/python3.7/site-packages/selenium/webdriver/remote/webelement.py", line 685, in find_elements
    {"using": by, "value": value})['value']
  File "/root/.local/lib/python3.7/site-packages/selenium/webdriver/remote/webelement.py", line 633, in _execute
    return self._parent.execute(command, params)
  File "/root/.local/lib/python3.7/site-packages/selenium/webdriver/remote/webdriver.py", line 321, in execute
    self.error_handler.check_response(response)
  File "/root/.local/lib/python3.7/site-packages/selenium/webdriver/remote/errorhandler.py", line 242, in check_response
    raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.StaleElementReferenceException: Message: The element reference of <main class="SCxLW uzKWK "> is stale; either the element is no longer attached to the DOM, it is not in the current frame context, or the document has been refreshed
```

**Comment by turtlebking** [comment]
This is the 'Follow' button on instagram html.

`<button class="_5f5mN       jIbKX  _6VtSN     yZn4P   ">Follow</button></span>`

may be a soft block?

**Comment by Jostino** [comment]
> This is the 'Follow' button on instagram html.
> 
> `<button class="_5f5mN jIbKX _6VtSN yZn4P ">Follow</button></span>`
> 
> may be a soft block?

I've logged in on the browser with the account with I reported the problem, but I'm able to Follow accounts. But the error was 6hrs ago actually.

**Comment by bradwalls** [comment]
> 
> This is the 'Follow' button on instagram html.
> 
> <button class="_5f5mN jIbKX _6VtSN yZn4P ">Follow</button></span>
> 
> may be a soft block?



I just ran quickstart without any following/unfollowing activity and still had the error on a seperate account.

**Comment by alexsmedile** [comment]
My settings are something like "follow 2% of the total account you interact with, wait 60 seconds" so I don't think it is a soft block. And if I do it manually I have no block on follow.

**Comment by juauzynhu** [comment]
I think problem is related with 'set_dont_unfollow_active_users'. I commented this line in my script and voilá

**Comment by alexsmedile** [comment]
> 
> 
> I think problem is related with 'set_dont_unfollow_active_users'. I commented this line in my script and voilá

I have this issue with the `follow_likers` function. Have no 'set_dont_unfollow_active_users' in it :(

**Comment by shomanishikawa** [comment]
Also seeing this issue. Not running anything related to `set_dont_unfollow_active_users`, though I do notice the error is coming from the `unfollow_util.py`

**Comment by Jostino** [comment]
> 
> 
> Also seeing this issue. Not running anything related to `set_dont_unfollow_active_users`, though I do notice the error is coming from the `unfollow_util.py`

I agree with the last statement cuz the allUnfollow didn't work anymore on the unfollow options...

**Comment by Jostino** [comment]
Update
i've tried a script that do only follows and and it still broken

```
Traceback (most recent call last):
  File "InstaPy/assets/vintage.py", line 48, in <module>
    sleep_delay=60)
  File "/root/.local/lib/python3.7/site-packages/instapy/instapy.py", line 3520, in follow_user_followers
    person, self.follow_times, sleep_delay, interact
  File "/root/.local/lib/python3.7/site-packages/instapy/instapy.py", line 1193, in follow_by_list
    self.logfolder,
  File "/root/.local/lib/python3.7/site-packages/instapy/unfollow_util.py", line 780, in follow_user
    browser, track, login, user_name, None, logger, logfolder
  File "/root/.local/lib/python3.7/site-packages/instapy/unfollow_util.py", line 211, in get_following_status
    following_status = follow_button.text
  File "/root/.local/lib/python3.7/site-packages/selenium/webdriver/remote/webelement.py", line 76, in text
    return self._execute(Command.GET_ELEMENT_TEXT)['value']
  File "/root/.local/lib/python3.7/site-packages/selenium/webdriver/remote/webelement.py", line 633, in _execute
    return self._parent.execute(command, params)
  File "/root/.local/lib/python3.7/site-packages/selenium/webdriver/remote/webdriver.py", line 321, in execute
    self.error_handler.check_response(response)
  File "/root/.local/lib/python3.7/site-packages/selenium/webdriver/remote/errorhandler.py", line 242, in check_response
    raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.StaleElementReferenceException: Message: The element reference of <button class="_5f5mN       jIbKX  _6VtSN     yZn4P   "> is stale; either the element is no longer attached to the DOM, it is not in the current frame context, or the document has been refreshed
```

I hope someone can find a solution soon :(

RaspBerry Pi 3+

**Comment by bradwalls** [comment]
This is still a showstopper for me, tried every interaction function (like, follow, unfollow, story) and returns the same error even with deleting cookie everytime

**Comment by meth0xy** [comment]
Same on Windows!

**Comment by flygohr** [comment]
Same here!

**Comment by bradwalls** [comment]
Is there any updates for this?

**Comment by ljacobo** [comment]
Tried with the DevBranch and the problem persists...

**Comment by turtlebking** [comment]
by looking at IG html
**Outer HTML From IG profile page:**

Following
```
<button class="_5f5mN    -fzfL     _6VtSN     yZn4P   ">Following</button>
or
<button class="_5f5mN    -fzfL     _6VtSN     yZn4P   "><div class="                  Igw0E   rBNOH          YBx95       _4EzTm                                                                                                              " style="height: 26px;"><span class="glyphsSpriteFriend_Follow u-__7" aria-label="Following"></span></div></button>
```
Follow
```
<button class="_5f5mN       jIbKX  _6VtSN     yZn4P   ">Follow</button>    <----this
```
---
the script broken because the 'Follow' button on user's profile page can not be located.

in xpath_complie.py at line 80
```
xpath["get_following_status"] = {
    "follow_button_XP": "//button[text()='Following' or \
                                  text()='Requested' or \
                                  text()='Follow' or \
                                  text()='Follow Back' or \
                                  text()='Unblock']",
    "follow_span_XP_following": "//button/div/span[contains(@aria-label, 'Following')]",
```
it should able to get the 'Follow/Following/Follow Back/Requested/Unblock' button status by button text. Perhaps your browser lang is not English? How often did the error come up?

**Comment by Jostino** [comment]
On my case i've installed firefox-esr from CLI, so I thought was already in English, there is a way to check it out? Someone known this way?

**Comment by turtlebking** [comment]
> On my case i've installed firefox-esr from CLI, so I thought was already in English, there is a way to check it out? Someone known this way?

you can check the follow button text on your IG page if it is in english

**Comment by Jostino** [comment]
> > On my case i've installed firefox-esr from CLI, so I thought was already in English, there is a way to check it out? Someone known this way?
> 
> you can check the follow button text on your IG page if it is in english

Well, I have lot of accounts, but if they use the the app in Italian Language is it related to the error here?

**Comment by turtlebking** [comment]
> > > On my case i've installed firefox-esr from CLI, so I thought was already in English, there is a way to check it out? Someone known this way?
> > 
> > 
> > you can check the follow button text on your IG page if it is in english
> 
> Well, I have lot of accounts, but if they use the the app in Italian Language is it related to the error here?

As a matter of fact, yes. the program only look up button text in English.

**Comment by Jostino** [comment]
This is a problem, how can I ask to all these accounts to set the language to english and why before Firefox this wasnt occured?
This look weird to me. Is it possible to InstaPy to force english language while doing the operations?

**Comment by Jostino** [comment]
> > > > On my case i've installed firefox-esr from CLI, so I thought was already in English, there is a way to check it out? Someone known this way?
> > > 
> > > 
> > > you can check the follow button text on your IG page if it is in english
> > 
> > 
> > Well, I have lot of accounts, but if they use the the app in Italian Language is it related to the error here?
> 
> As a matter of fact, yes. the program only look up button text in English.

This is a problem, how can I ask to all these accounts to set the language to english and why before Firefox this wasnt occured?
This look weird to me. Is it possible to InstaPy to force english language while doing the operations?

**Comment by turtlebking** [comment]
i will see if i can update xpath_complie.py to fit any lang

**Comment by rabb1t-dev** [comment]
I have this exact same issue and my browser language is set to English. For me the error happens within a half hour of starting my script, usually within ten minutes.

**Comment by turtlebking** [comment]
In browser.py, Firefox is already set to English.
But there is some bugs there.

**Comment by turtlebking** [comment]
edit browser.py start from line 88-131

```
    if browser_profile_path is not None:
        profile = webdriver.FirefoxProfile(browser_profile_path)
    else:
        profile = webdriver.FirefoxProfile()

    # set English language
    profile.set_preference("intl.accept_languages", "en-US")
    profile.set_preference("general.useragent.override", user_agent)

    if disable_image_load:
        # permissions.default.image = 2: Disable images load,
        # this setting can improve pageload & save bandwidth
        profile.set_preference("permissions.default.image", 2)

    if proxy_address and proxy_port:
        profile.set_preference("network.proxy.type", 1)
        profile.set_preference("network.proxy.http", proxy_address)
        profile.set_preference("network.proxy.http_port", proxy_port)
        profile.set_preference("network.proxy.ssl", proxy_address)
        profile.set_preference("network.proxy.ssl_port", proxy_port)

    # mute audio while watching stories
    profile.set_preference("media.volume_scale", "0.0")

    # prefer user path before downloaded one
    driver_path = geckodriver_path or get_geckodriver()
    try:
        browser = webdriver.Firefox(
            firefox_profile=profile,
            executable_path=driver_path,
            options=firefox_options,
        )
    except SessionNotCreatedException:
        print("Unable to locate the default Firefox binary path, trying to find it...")
        firefox_binary_path = get_firefox_binary_path()
        if firefox_binary_path is None:
            print("Unable to start Firefox, is it installed?")
        else:
            browser = webdriver.Firefox(
                firefox_binary=firefox_binary_path,
                firefox_profile=profile,
                executable_path=driver_path,
                options=firefox_options,
            )

```

**Comment by Jostino** [comment]
> edit browser.py start from line 88-131
> 
> ```
>     if browser_profile_path is not None:
>         profile = webdriver.FirefoxProfile(browser_profile_path)
>     else:
>         profile = webdriver.FirefoxProfile()
> 
>     # set English language
>     profile.set_preference("intl.accept_languages", "en-US")
>     profile.set_preference("general.useragent.override", user_agent)
> 
>     if disable_image_load:
>         # permissions.default.image = 2: Disable images load,
>         # this setting can improve pageload & save bandwidth
>         profile.set_preference("permissions.default.image", 2)
> 
>     if proxy_address and proxy_port:
>         profile.set_preference("network.proxy.type", 1)
>         profile.set_preference("network.proxy.http", proxy_address)
>         profile.set_preference("network.proxy.http_port", proxy_port)
>         profile.set_preference("network.proxy.ssl", proxy_address)
>         profile.set_preference("network.proxy.ssl_port", proxy_port)
> 
>     # mute audio while watching stories
>     profile.set_preference("media.volume_scale", "0.0")
> 
>     # prefer user path before downloaded one
>     driver_path = geckodriver_path or get_geckodriver()
>     try:
>         browser = webdriver.Firefox(
>             firefox_profile=profile,
>             executable_path=driver_path,
>             options=firefox_options,
>         )
>     except SessionNotCreatedException:
>         print("Unable to locate the default Firefox binary path, trying to find it...")
>         firefox_binary_path = get_firefox_binary_path()
>         if firefox_binary_path is None:
>             print("Unable to start Firefox, is it installed?")
>         else:
>             browser = webdriver.Firefox(
>                 firefox_binary=firefox_binary_path,
>                 firefox_profile=profile,
>                 executable_path=driver_path,
>                 options=firefox_options,
>             )
> ```

Thanks, i've done the changes and the previous script now show me another error related to another issue i've already opened, so to me this problem is not reported now.

**Comment by riccardovalente** [comment]
I still get the error above occasionally, even with the changes you suggested.

**Comment by turtlebking** [comment]
if you are using Raspberry Pi 3+, try geckodriver 0.24.0 arm7hf (unzip it)
put it in \usr\local\bin
[geckodriver.zip](https://github.com/timgrossmann/InstaPy/files/3627590/geckodriver.zip)

**Comment by riccardovalente** [comment]
I have been using that for a while already.

**Comment by FeatherKamisori** [comment]
Same issue on windows 10, note that first firefox crashes (disappears/closes).

**Comment by papillon-unix** [comment]
@turtlebking I just tried your edit on Instapy 0.6.4. on Raspberry PI 3 (with geckodriver 0.24.0) and it worked for **only one follow**, then crashed again with the same error..

**Comment by Jostino** [comment]
> @turtlebking I just tried your edit on Instapy 0.6.4. on Raspberry PI 3 (with geckodriver 0.24.0) and it worked for **only one follow**, then crashed again with the same error..

Same instapy version(but dev branch), same Raspberry, same geckodriver version, firefox-esr 60
same issue

**Comment by papillon-unix** [comment]
> Same instapy version(but dev branch), same Raspberry, same geckodriver version, firefox-esr 60
> same issue

I just tried on dev branch as well but same issue..

**Comment by rabb1t-dev** [comment]
Anyone found a fix for this? InstaPy is completely unusable for me...

**Comment by papillon-unix** [comment]
@rabb1t0x21 Yes: #5068 ! 
It wasn't easy to find but someone on Discord had a raspberry working good so he told me his exact config :
- Instapy 0.6.4 dev branch
- firefox-esr 60
- geckodriver 0.24.0
- and then this [manual fix](https://github.com/timgrossmann/InstaPy/issues/5068#issuecomment-534010224)

It worked for me too, but from now on I run Instapy extremly slow to avoid being flagged, because these bugs are all about Instagram turning really tough with bots.

**Comment by rabb1t-dev** [comment]
Wow, actually I can confirm that @papillon-unix solutions has worked for me. InstaPy ran all night long, still going with no issues on Ubuntu 18.04 running headless=False.

**Comment by papillon-unix** [comment]
In fact, I made a mistake in my post, I run `geckodriver 0.23.0` and not 0.24.
0.24 was crashing Instapy.

**Comment by hamzadiaz** [comment]
I have the same error, is there any solution??

**Comment by khera-shanu** [comment]
Yup this issue is still there!

**Comment by matthewblake101** [comment]
I found that this is due to the script trying to find the button before the page is loaded properly. Thus, I found a workaround - go to line 212 of Unfollow_util.py.

There is:  following_status = follow_button.text

I let it try and if it fails then I wait, refresh the page and then try again by calling the function.

    try:
        following_status = follow_button.text  #this is what normally fails
    except:
        reload_webpage(browser)
        sleep(5)
        following_status, follow_button = get_following_status(browser, track, username, person, person_id, logger, logfolder)

**Comment by chuddyni** [comment]
> I found that this is due to the script trying to find the button before the page is loaded properly. Thus, I found a workaround - go to line 212 of Unfollow_util.py.
> 
> There is: following_status = follow_button.text
> 
> I let it try and if it fails then I wait, refresh the page and then try again by calling the function.
> 
> ```
> try:
>     following_status = follow_button.text  #this is what normally fails
> except:
>     reload_webpage(browser)
>     sleep(5)
>     following_status, follow_button = get_following_status(browser, track, username, person, person_id, logger, logfolder)
> ```
unfollow_util.py is moved to follow_utils.py atm
